### ISSUE TITLE
Allow clear_comtypes_cache to be invoked as e.g. py -m comtypes.clear_cache

### ISSUE THREAD

**Issue by bennyrowland** [enthought/comtypes#477]
I would like to be able to invoke the `clear_comtypes_cache` script from another script easily on a heterogeneous set of installed systems, but currently finding its location is not trivial. Some of my systems use the primary Python installation while others use `venv`s, so the relative path from the python executable is not consistent and I don't particularly want to write a helper script that will hunt around for its location. It seems to me that a neater solution would be to allow the script to be invoked as a module, then as long as the Python launcher is going to find the correct environment then I can just run something like: `py -m comtypes.clear_cache` and know that it will work reliably across different systems and different Python installations on a single system.

I think that this would be a simple change to make, basically just moving the existing `clear_comtypes_cache.py` file inside the comtypes source, potentially renaming it to (since the package specifies it is `comtypes` I would suggest dropping that from the module name but not committed to that), and then modifying `setup.py` to create the `clear_comtypes_cache" script entry using a `console_scripts` entry point. If the maintainers are ok with the change, I am happy to put together the PR.

**Comment by junkmd** [comment]
I think your suggestions are excellent.

In my project using `comtypes`, I wrote a project-specific process to remove "Partial wrapper files generated in gen folder" as shown in #116.

If such a process could be called from `comtypes`, it would remove from projects a program that has taken on excessive responsibility.

What I am concerned about is the testing.
I don't want the `comtypes/gen` folder of the executing environment to be actually deleted when the test suite is run.

I think we need `unittest.mock` to ease testing.
The `master` branch supports Python2 that has no `unittest.mock`.
I propose a PR to the `drop_py2` branch, which supports for Python3 only.

**Comment by bennyrowland** [comment]
I have an implementation of this working locally, but it uses the 3rd party package pyfakefs which provides an in-memory file system. This allows comtypes to see the real gen folder, but when it is deleted it is only in the temporary overlay version rather than the real file system. Is adding a third party package for testing purposes only acceptable? It would be possible to do this with unittest.mock instead but it would be a lot more complex to match what pyfakefs gives us. A simple option would be just to mock out the `shutil.rmtree()` function that is used to actually clear the cache folder, but this is rather more brittle because if that function was changed the test would break even if the script was still correct (and it would also allow the test to delete the real files in that case). @junkmd what do you think?

**Comment by junkmd** [comment]
I think introducing `pyfakefs` for this test is a bit much.
It would require changing the CI workflow and/or modifying the documentation about the test.
Rather, I would prefer to use `unittest.mock`.

The sample test codebase below depends on the production codebase implementation.

```py
# production.py
from pathlib import Path
from shutil import rmtree as _rmtree  # THE TEST RELIES ON THIS FUNCTION BEING IMPORTED WITH THIS NAME
from typing import Path


def get_next_cache_dir() -> Optional[Path]:
    ...


def main():
    ...


if __name__ == "__main__":
    main()
```

```py
# test_foo.py
import unittest as ut
from unittest import mock
import foo


class Test_get_next_cache_dir(ut.TestCase):
    ...


class Test_main(ut.TestCase):
    def setUp(self):
        self.patched_rmtree = mock.patch.object(foo, "_rmtree")
        self.patched_rmtree.start()
        # "If setUp() fails, meaning that tearDown() is not called,
        # then any cleanup functions added will still be called."
        # https://docs.python.org/3/library/unittest.html#unittest.TestCase.addCleanup
        self.addCleanup(self.patched_rmtree.stop)
        self.patched_get_next_cache_dir = mock.patch.object(foo, "get_next_cache_dir")
        self.patched_get_next_cache_dir.start()
        self.addCleanup(self.patched_get_next_cache_dir.stop)

    def test_spam(self):
        self.patched_get_next_cache_dir.side_effect = [...]
        foo.main()
        self.assertEqual(self.patched_rmtree.call_args_list, [...])
    # and tests for other conditions...
```

This test is brittle if the production codebase changes often.

However, [the `clear_comtypes_cache.py` commit log](https://github.com/enthought/comtypes/commits/d186dca/clear_comtypes_cache.py) shows that is changed less frequently than the other files.
And I can't think of any situation where new functionality will be added to this module after this change.

In other words, I think "the cost of introducing a new third-party package for testing" is more expensive than "the cost of writing tests for a production codebase that is likely to change less frequently" for the `clear_comtypes_cache.py` test.

Of course, no one knows what the future holds, so there is **no** such thing as "we will never consider a new third party testing package forever".
However, with this issue, the idea is that we can get by without introducing it yet.

**Comment by bennyrowland** [comment]
That is fair enough for now, I have submitted a PR with a version without 3rd party packages. However, I expect I will also raise another issue sometime soon to explore updating the tooling for comtypes - things like tox/pre-commit/ruff etc. would make a lot of this stuff easier to implement.
